File: tests/poo_002/solver.ts

// Nesse rascunho esta faltando a parte de dirigir o carro

//++1

class Car{
    pass: number; // Passageiros
    passMax: number; // limite de Passageiros
    gas: number; // tanque
    gasMax: number; // limite do tanque
    km: number; // quantidade de quilometragem
//++1

    constructor() {
        this.pass = 0; // Passageiros
        this.passMax = 2; // limite de Passageiros
        this.gas = 0; // tanque
        this.gasMax = 100; // limite do tanque
        this.km = 0; // quantidade de quilometragem
    }

    enter(): void {
        if (this.pass < this.passMax) {
            this.pass += 1;
            return;
        }
        write("fail: limite de pessoas atingido");
    }

    leave(): void {
        if (this.pass > 0) {
            this.pass-=1;
            return;
        }
        write("fail: nao ha ninguem no carro");
    }

    fuel(gas: number): void {
        this.gas += gas;
        if(this.gas > this.gasMax)
            this.gas = this.gasMax;
    }
    
//--
    drive (km: number): void {
        if(this.pass == 0) {
            write("fail: nao ha ninguem no carro");
        } else if(this.gas == 0) {
            write("fail: tanque vazio");
        }
        else if(this.gas < km) {
            write("fail: tanque vazio apos andar " + this.gas + " km");
            this.km += this.gas;
            this.gas = 0;
        } else{
            this.gas = this.gas - km;
            this.km = this.km + km;
        }
    }

//++1
    toString(): string {
        return "pass: " + this.pass + ", gas: " + this.gas + ", km: " + this.km;
    }
};

//==

let _cin_ : string[] = [];
try { _cin_ = require("fs").readFileSync(0).toString().split(/\r?\n/); } catch(e){}
let input = () : string => _cin_.length === 0 ? "" : _cin_.shift()!;
let write = (text: any, end:string="\n")=> process.stdout.write("" + text + end);

function main() {
    let car = new Car();

    while (true) {
        let line = input();
        write("$" + line);
        let args = line.split(" ");

        if      (args[0] === "show")  { write(car.toString());          }
        else if (args[0] === "enter") { car.enter();                    }
        else if (args[0] === "leave") { car.leave();                    }
        else if (args[0] === "fuel")  { car.fuel(+args[1]);             }
//--
        else if (args[0] === "drive") { car.drive(+args[1]);            }
//==
        else if (args[0] === "end")   { break;                          }
        else                          { write("fail: comando invalido");}
    }
}

main()

File: tests/03_run/poo_002/solver.cpp
#include "fn.hpp" // https://raw.githubusercontent.com/senapk/cppaux/master/fn.hpp

// Nesse rascunho esta faltando a parte de dirigir o carro

//++1
class Car {
public:
    int pass;
    int passMax;
    int gas;
    int gasMax;
    int km;
//++1

    Car() {
        pass = 0;
        passMax = 2;
        gas = 0;
        gasMax = 100;
        km = 0;
    }

    void enter() {
        if (pass < passMax) {
            pass++;
        } else {
            fn::write("fail: limite de pessoas atingido");
        }
    }

    void leave() {
        if (pass > 0) {
            pass--;
        } else {
            fn::write("fail: nao ha ninguem no carro");
        }
    }

    void fuel(int gas) {
        this->gas += gas;
        if (this->gas > gasMax) {
            this->gas = gasMax;
        }
    }
//--
    void drive(int km) {
        if (pass == 0) {
            fn::write("fail: nao ha ninguem no carro");
        } else if (gas == 0) {
            fn::write("fail: tanque vazio");
        } else if (this->gas < km) {
            fn::print("fail: tanque vazio apos andar {} km\n", this->gas);
            this->km += this->gas;
            this->gas = 0;
        } else {
            this->km += km;
            this->gas -= km;
        }
    }
    
//++1
    std::string str() {
        return fn::format("pass: {}, gas: {}, km: {}", pass, gas, km);
    }
};

//==
int main() {
    Car car;
    while (true) {
        auto line = fn::input();
        auto args = fn::split(line, ' ');
        fn::write("$" + line);

        if      (args[0] == "show")  { fn::write(car.str());                }
        else if (args[0] == "enter") { car.enter();                         } 
        else if (args[0] == "leave") { car.leave();                         }
        else if (args[0] == "fuel")  { car.fuel(+args[1]);                  }
//--
        else if (args[0] == "drive") { car.drive(+args[1]);                 }
//==
        else if (args[0] == "end")   { break;                               }
        else                         { fn::write("fail: comando invalido"); }
    }
}
